File: problems/020-offline-dependency-bundle/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type R struct {
	d string
	z uint64
}

func find(a []R, d string) int {
	i := sort.Search(len(a), func(i int) bool { return a[i].d >= d })
	if i < len(a) && a[i].d == d {
		return i
	}
	return -1
}
func main() {
	in := bufio.NewReader(os.Stdin)
	var n, m int
	fmt.Fscan(in, &n, &m)
	l := make([]R, n)
	p := make([]R, m)
	var total uint64
	for i := range l {
		var name string
		fmt.Fscan(in, &name, &l[i].d, &l[i].z)
		total += l[i].z
	}
	for i := range p {
		fmt.Fscan(in, &p[i].d, &p[i].z)
	}
	sort.Slice(l, func(i, j int) bool { return l[i].d < l[j].d })
	sort.Slice(p, func(i, j int) bool { return p[i].d < p[j].d })
	req := []R{}
	for i := 0; i < n; {
		j := i + 1
		for j < n && l[j].d == l[i].d {
			if l[j].z != l[i].z {
				fmt.Println("LOCK_CONFLICT", l[i].d)
				return
			}
			j++
		}
		req = append(req, l[i])
		i = j
	}
	for i := 1; i < m; i++ {
		if p[i].d == p[i-1].d {
			fmt.Println("DUPLICATE_PAYLOAD", p[i].d)
			return
		}
	}
	for _, x := range req {
		if find(p, x.d) < 0 {
			fmt.Println("MISSING", x.d)
			return
		}
	}
	for _, x := range p {
		if find(req, x.d) < 0 {
			fmt.Println("EXTRA", x.d)
			return
		}
	}
	var unique uint64
	for _, x := range req {
		if p[find(p, x.d)].z != x.z {
			fmt.Println("SIZE", x.d)
			return
		}
		unique += x.z
	}
	fmt.Println("VALID", len(req), unique, total-unique)
}

File: problems/020-offline-dependency-bundle/solutions/rust/main.rs
use std::io::{self, Read};
fn pos(a: &[(String, u64)], d: &str) -> Result<usize, usize> {
    a.binary_search_by(|x| x.0.as_str().cmp(d))
}
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut it = s.split_whitespace();
    let n: usize = it.next().unwrap().parse().unwrap();
    let m: usize = it.next().unwrap().parse().unwrap();
    let mut l = vec![];
    let mut total = 0u64;
    for _ in 0..n {
        it.next();
        let d = it.next().unwrap().to_string();
        let z = it.next().unwrap().parse().unwrap();
        total += z;
        l.push((d, z));
    }
    let mut p: Vec<(String, u64)> = (0..m)
        .map(|_| {
            (
                it.next().unwrap().to_string(),
                it.next().unwrap().parse().unwrap(),
            )
        })
        .collect();
    l.sort();
    p.sort();
    let mut req = vec![];
    let mut i = 0;
    while i < n {
        let mut j = i + 1;
        while j < n && l[j].0 == l[i].0 {
            if l[j].1 != l[i].1 {
                println!("LOCK_CONFLICT {}", l[i].0);
                return;
            }
            j += 1
        }
        req.push(l[i].clone());
        i = j
    }
    for i in 1..m {
        if p[i].0 == p[i - 1].0 {
            println!("DUPLICATE_PAYLOAD {}", p[i].0);
            return;
        }
    }
    for x in &req {
        if pos(&p, &x.0).is_err() {
            println!("MISSING {}", x.0);
            return;
        }
    }
    for x in &p {
        if pos(&req, &x.0).is_err() {
            println!("EXTRA {}", x.0);
            return;
        }
    }
    let mut unique = 0;
    for x in &req {
        let y = &p[pos(&p, &x.0).unwrap()];
        if x.1 != y.1 {
            println!("SIZE {}", x.0);
            return;
        }
        unique += x.1
    }
    println!("VALID {} {} {}", req.len(), unique, total - unique);
}
